### ISSUE TITLE
Experience report upgrading to 4.5.0

### ISSUE THREAD

**Issue by pbiggar** [fsprojects/fantomas#1854]
Hey, I felt like this might be useful insight, feel free to close this issue or whatever.

I spent a few hours yesterday upgrading to the latest fantomas. The PR is here: https://github.com/darklang/dark/pull/3153/files

# IDE integration
We had been having an issue for months that files would fail or formatting. This is because we had Fantomas 4.4.0 installed in our devcontainer and that was used for CI checks, but VSCode automatically (and really quite aggressively) upgrades people to newer versions of Ionide whenever one is released (even if we previously chose to stay on an older version 🤬) which upgrades Fantomas which causes formatting errors. This is kinda the problem that formatters are supposed to avoid, so this was quite frustrating. It seems like you've got this on your radar with #1844, big +1 for reusing the installed version.

# Indentation change

The major change across the codebase was that the depth of the indent was changed in pattern matches. It previously indented and now it does not.

I think the horizontal space saving is great here, even if my eyes are a little used to seeing the indent (I see quite a few places where things fit on one line, which is always nice, [egregious example here](https://github.com/darklang/dark/pull/3153/files?diff=split&w=1#diff-93ecf1065bafe433dc533f950bd275bce7c688f84afcd4d4f6ed1320a8a6be24)).

That said, it is quite frustrating to have significant changes like this. I'm trying to manually review the changes and they're so big it's very challenging. GitHub refuses to show many of the diffs as they're too big.  It seems that this didn't come with a configuration flag either - I might have disabled it to start so I could check other changes and reenabled it in a second commit.

# TryWith change
This change I really don't like:

Before:
```
with e ->
```

After:
```
      with
      | e ->
```

This is a really unpleasant change that had me immediately looking for a flag to disable it - there does not appear to be one.

# `when ... -> ...` change

It seems that when a an expression body in a pattern match fits on the same line as a `when`, fantomas will now move it to a new line (including giving `->` a line of its own). This is ugly imo, and seems like a bug (certainly giving the arrow its own line must be a bug?): https://github.com/darklang/dark/pull/3153/files?diff=split&w=1#diff-e0e2e1ab4264a523e4a703183857428b285615132d22ddb2a86a9c4ee3a94989R275. I see this in many places in my codebase now.

# SQL formatting
One issue from the upgrade is this sort of change to our SQL code: https://github.com/darklang/dark/pull/3153/files#diff-93ecf1065bafe433dc533f950bd275bce7c688f84afcd4d4f6ed1320a8a6be24.

Long strings seem to now start on a newline. I'm unsure of my feelings here in the general case though I think it might be better this way. In the SQL case, though, we had manually lined up the strings for readability, which are now no longer lined up. and it didn't fix them (of course it can't fix them as that whitespace could be significant). 

# Overall

Overall seems some nice improvements, looking forward to the editor integration being consistent.

Will I file bugs for the TryWith and when-> changes?

**Comment by nojaf** [comment]
Hey Paul,

Thanks for this report.

#### IDE integration

Yes, this is a pain point targeted for the next release. The reason why Ionide is always updating Fantomas is because of the shared [FCS](https://fsharp.github.io/FSharp.Compiler.Service/) dependency. When a new version of that drops, a bunch of NuGet packages need to bump at the same time. We want to decouple this dependency by formatting out of process.

#### Indentation change
The current indentation rules for F# are a bit weird when it comes to `match/with` and `if/then` expressions. This is being fixed at [the compiler side](https://github.com/dotnet/fsharp/pull/11772) and we should in the future have a more consistent story there.

#### TryWith

This was updated due to new guidance from the two style guides.

> Pattern matching on the exception type should be indented at the same level as with.

[MS guide](https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#formatting-trywith-expressions).
[GR guide](https://github.com/G-Research/fsharp-formatting-conventions#formatting-trywith-expressions)
 
I've started to look at `try/with` the same way you would write a `match/with`.
If the match expression is horribly long (which I would always avoid) you would place it between the `match` and `with` keywords with one indent.

```fsharp
match 
    // something really long
    longExpr
with
| _ -> ()

try
   // something long as well but generally more accepted.
  longExpr
with
| ex -> ()
```

#### SQL formatting

Any examples not in a match clause? That really was the indentation change.
Otherwise, strings are just treated as any regular argument would be in a multiline function application.

#### Overall

> Will I file bugs for the TryWith and when-> changes?

This actually works as expected. [Example](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAbRAXMAzb0AFgMIDyASgMqXZYCWAdgOZgAU0AhgwG4cDOASQAiaYl16ChASjBoAKvwDWAHlIAHerAbLisAE58aWeswB8psAF4AOgwiUAjugB0DgK6I9ATzDXIlAFEAGQDiOTA4AwB9PlpGFgAxclIAWQjxfkQ%2BMAB1AAkA8gCwOngrMAABTh5%2BYT9bCAAfC0cXNQ49DgBbbE9sgG1fGAzJPwAaMFbnNzdS9JrJMABdBrBmyadnRAAPRGg3LERyWAB3AEE%2BLwZoVYg71lw3Oz1EDjKAWlNbu7uujiwiGAXm9nHwsHp4qQ9AA5LSIIaRPgxOLMPxgE50LCEb4-NZgWEMeGffFwnE-RqTWA9MBgvRgT5k3GYHAAKz4WnKlC8YMQXWcch2WGcACl2QwRWKhLB9j0GEKAAodWI08G2Rk-P4AwhgNlaZzHWBYAKYWVCgBqHHQHgA0owyhiseq7gB6Z1gLF0bLoRjwz1gACcAGYIoQOhxoIcDGB0FomE6mpNuYc%2BQLtkLRXqMwwLVbELaGPBnJRwfF0YREHZdeKDUaTRWhQBxbDFiHMVgySxDABUaIZdlxA8p1LO6HQULoTEYfHjeK5PJTgolmbFOZtducZz0nR8fcHA6r%2BtghuNvPrzgCDDcPU6h032-bM6aLUQDmcfzUbEediJFkQ-NgLbxO2UiPmsz6vlgsBBJ6WCges46Tgw079nudzrJQVKIDOFJRPSFi4BwdDoA62p%2BIwvDemUVYlHYugGEYJhMNOkAgQwtiQGM4BQLExjxHwqBgP0XEQKAA6QAAJFgXhqMgaCQIwWAccJdwSRRHgCUJe4AAycXukACAW9aUHQABeyC6YOABMykQEsykAL4WSJNlQJJ0myVAClKWJ4lqR5mmDgAjE5uKQCkHDbNBhJBBWTBYt5g4ABwAKwuXZPyOcpomhW5MkCZAABGR7oAlPyqZa6loAFA6WSFZU0F0dBwDGDBnEaBakLgUXmS5EAEegsRpQ5IXZWVuUeYVxWlSpvkVf5vVgIGdUqZQ7TQIgABCiC4PoiCKp0PSRtNPzgh4Q0ZSNLkSVJeVyUVsAlct5W5hpC0ACzLRA-hrZt227UEpyeJwsQGdw0r-HQWjHXcp2IOddyZT8o0zTdE33Y9V2zS9VULcln1QKt4a-TtLwAKpqDJejA4goPg5oDDQ31lqDQO6UI5dPmo-l6PQ89lWCQtABs%2BPfUTW0k4gxDoPwfC6Ehp0RvojN4MzcOs8NWWY1zd1TU9WP89VuIAOwi4T63i7tKS8gVnjK-1LO4mzECI3cyNfeN3O65jfmvXuiWmz9FsvLoLV26r8POxzOXa1APN6z7ON7v6AdE2cuCRroXR-MrsMR2ALvOZz7mew9vP6-Ne6BTpC2i%2Bbf0vA1TUPVD%2BP22rjsa0jWvFzrpfx3NvtBcFNdm4gaeRo3zUtwtufqxdmtF7dsdez5CcC5XtU1wZSByqQDByN4OSYoQYcDe3PxO-nUdjTHk1997A%2BJ0FS0jz9m6wE88BCIg3qNUd%2BOzw7vPLui8JpeX7tjdeQUPo13CtsAQuA5DlgYMaWIlBCD6CwEfeA8V8ZvS0nnAuYA3auVvuAh%2BkDDY-ECnjWBEUDK4DoNsdQnh-j6ACNsNQLw%2BB8EhgzfGyUCFz3ZgvaOPdPJyjLmvKhdxArCzodscgex9DwGwbg96QigEiJAWIpe8lJEQINgtQKJsFFKMiPAaEV4bZ6E6gIZM-F8aBUIdfFG4jIBdDcOgeg3pCQJH0JqSMchi6GIrkFf2NdzEqJSF4nxPp-F6ECbbEWRAwwRk8FEDEOCT4uNETfdx5DV6PygQOQKycFF3g4F4KE0EwRqJPngzRF9O6u27nowpOVpELUstXPScDKnVL0LUrAViug2LsQ45WzjhGRzyW4vRnjvFEXiQE-4QSQkUKMXuSyw89IDJqTBGJSzfGIASUkvQysYChk6OkvQmTSjxVyTo-J7SDGbLCTVTefSIqrk2naeI9Tlb4Kea00B%2BUOljS6dsl%2B3ztgJCeBGPhG1-nMEBY0kFhddFgLeUUyh3SYGwqlFgJsRpOHcN4VoNFuMml3EvkQkh113Fx3eYPGqtC%2BmxOWYSDaMZoCKA2jcxQ2A%2BB70oN0SWD0rz8IWm3DFxC2loxXp04pMiICWXkXpaEiATgnK2lgE4iAKzBJkt-RhDBMR8LOAWK2Yy%2Binwds04BoKsUlwxrirZVlTF6WtIatQCCkEVgMmKno3V7Xn1pS0zFLzsWKVCay3ElkImwuNI1PghAqV7mBTMq%2Bcz3a32Ze6j5Cbyl6WMswTAm4mBXnrDkRABUUgIB6nuWV2b6UKtdVIlVC1Ay9MHJAEcE4GDwuuPTYyTAGD-DcC8OQsBt71ghtPZt4dW2uLzUypVkKu17kDLsvtI5IwTvoNwRAANmA2psaaxgFqtCOJlcurRsznnzMVffQt8afiBi%2BX2o59BupBG6AVeAHApawF4cwLVOqfRhrlQyj2vc3XKrxdumFfahCeg4AVTAKbPSEC5HKCK0GV25tIe4is8BOrdWLF4TAnakODkDASvtF5yNdSgykvQ6BcCQBg%2B2%2BDtGPUDkDOyvtPrEB%2BsMnKAyAquBEEIw%2BnNT6116ILYhgTuJAwar7Tyrgihup8Hfp-LVPJ4A-q5YgDhXCsgUqQjnPQZ0iOKZI8pjdM0oX0a9VpjoQdEBob4NACEjVD2IHgKTc1Whv7QGljePhcnHXaOddGjtcan6CaTX2wCEYG1IFixGjKXElgWXknwBIvCBKyvskAA).
The reason why the arrow is on the next line is probably related to another case where [the arrow after the expression leads to invalid code](https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWQC2VALswAswADzAB3RLwEAdZAB8wATTABPcQNjIwsiBB78haiVJ26FALTAAvMAFoAfNbABeMADZTEBQH07j6nSwYAD0wX6e4Vq6ABQAlJAgAL5AA). 
In short, the arrow on the next line always works, so that is why I went with it.
What you could do is open an issue on the [F# style guide](https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#feedback) and ask for guidance in that scenario.
![image](https://user-images.githubusercontent.com/2621499/128473422-4faa5115-2119-4395-a459-a0d8f625573c.png)

Thanks again for this report and your continued usage in such a large code-base.

**Comment by dsyme** [comment]
> Pattern matching on the exception type should be indented at the same level as with.

Ugh, I'm really very surprised to see this in the MS guide. This code is totally fine, despite what the guide says
```
try
    persistState currentState
with ex ->
    printfn "Something went wrong: %A" ex
```

I think we should update the guide here - is there any history to the addition of that?

**Comment by dsyme** [comment]
See https://github.com/dotnet/docs/pull/25498 for possible change here - @nojaf apologies in advance for messing Fantomas around here, assuming we make the change....

**Comment by pbiggar** [comment]
> The reason why the arrow is on the next line is probably related to another case where the arrow after the expression leads to invalid code.
> In short, the arrow on the next line always works, so that is why I went with it.


This took me a while to understand so let me restate what you said to make sure we're on the same page:

If we put the arrow on the same line, this code:

```
match x with
| Y y when 
    match y with
    | Z z -> z = 6
    | _ -> false // ->
    ->
    ()
```

would be intended as such:

```
match x with
| Y y when 
    match y with
    | Z z -> z = 6
    | _ -> false -> ()
```

which is syntactically invalid.

It would seem to me that this is quite an edge case, which must of course be kept working, but which I wouldn't expect to come up in practice.

I would suggest that we identify when this can happen and either a) only add a newline in that case or b) add parens and then move the arrow onto the same line, space permitting:

```
match x with
| Y y when 
    (match y with
    | Z z -> z = 6
    | _ -> false) -> ()
```


The change being discussed, where:

```
           match json.RootElement.ValueKind with
           | System.Text.Json.JsonValueKind.String when
               json.RootElement.GetString() = "*" -> Some AllOrigins
```

formats into:

```
           match json.RootElement.ValueKind with
           | System.Text.Json.JsonValueKind.String when
             json.RootElement.GetString() = "*"
             ->
             Some AllOrigins
```

To my eyes, this is a major eyesore, and the construct is relatively common. I think it's worth fixing this.

**Comment by nojaf** [comment]
Hey Paul, you understood correctly and we can follow up on this in another issue.
The thing is, the `when` expression can be over 100 different things (see [SynExpr](https://fsharp.github.io/FSharp.Compiler.Service/reference/fsharp-compiler-syntaxtree-synexpr.html)).
So, I'm pretty sure that it is not just `match` expressions but it will probably be other stuff as well.
I just went with something that would always produce valid syntax. I would go for an active pattern anyway when faced with such a situation. Please poke the style guide here as well.
Thanks for understanding.
